src/communicate.rs

```diff
@@ -145,96 +145,110 @@ pub fn write_to_peer_connection(
-                peer_addr,
-                &From::from(e),
-                "Open-Unidirectional",
-                user_msg,
-            )
-        })
-        .and_then(move |o_stream| {
-            let (message, msg_flag) = wire_msg.into();
+    let uni_stream = conn.open_uni();
 
-            tokio::io::write_all(o_stream, message)
-                .and_then(move |(o_stream, _)| tokio::io::write_all(o_stream, [msg_flag]))
-                .map_err(move |e| {
-                    utils::handle_communication_err(
-                        peer_addr,
-                        &From::from(e),
-                        "Write-All",
-                        user_msg0,
-                    )
-                })
-        })
-        .and_then(move |(o_stream, _): (_, [u8; 1])| {
-            tokio::io::shutdown(o_stream).map_err(move |e| {
+    let leaf = async move {
+        let mut o_stream = match uni_stream.await {
+            Ok(o_stream) => o_stream,
+            Err(e) => {
                 utils::handle_communication_err(
                     peer_addr,
                     &From::from(e),
-                    "Shutdown-after-write",
-                    user_msg1,
-                )
-            })
-        })
-        .map(move |_| utils::handle_send_success(peer_addr, user_msg2));
+                    "Open-Unidirectional",
+                    user_msg,
+                );
+                return;
+            }
+        };
 
-    current_thread::spawn(leaf);
-}
+        let (message, msg_flag) = wire_msg.into();
 
-/// Listen for incoming streams containing peer messages and read them when available
-pub fn read_from_peer(peer_addr: SocketAddr, incoming_streams: quinn::IncomingStreams) {
-    let leaf = incoming_streams
-        .map_err(move |e| {
+        if let Err(e) = o_stream.write_all(&message[..]).await {
+            utils::handle_communication_err(peer_addr, &From::from(e), "Write-All", user_msg);
+            return;
+        }
+        if let Err(e) = o_stream.write_all(&[msg_flag]).await {
+            utils::handle_communication_err(peer_addr, &From::from(e), "Write-All", user_msg);
+            return;
+        }
+
+        if let Err(e) = o_stream.finish().await {
             utils::handle_communication_err(
                 peer_addr,
                 &From::from(e),
-                "Incoming streams failed",
-                None,
+                "Shutdown-after-write",
+                user_msg,
             );
-        })
-        .for_each(move |quic_stream| {
-            read_peer_stream(peer_addr, quic_stream).map_err(|e| {
-                debug!(
-                    "Error in Incoming-streams while reading from peer {}: {:?} - {}.",
-                    peer_addr, e, e
-                )
-            })
-        });
+            return;
+        }
+
+        utils::handle_send_success(peer_addr, user_msg);
+    };
 
-    current_thread::spawn(leaf);
+    let _ = tokio::spawn(leaf);
```

Comment:
Is this _ pattern required for tokio?

Reply:
Yes, [`tokio::spawn`](https://docs.rs/tokio/0.2.8/tokio/fn.spawn.html) returns a `JoinHandle` now which is similar to a handle returned by `thread::spawn`. We don't have a use for it, hence `let _ = ...`